nixpkgs/toolchains/go.bzl

```diff
@@ -7,31 +7,244 @@ dependencies on rules_go for those who don't need go toolchain.
 `@io_tweag_rules_nixpkgs//nixpkgs:toolchains/go.bzl`.**
 """
 
-load(
-    "@io_bazel_rules_go//go:deps.bzl",
-    "go_wrap_sdk",
+load("//nixpkgs:nixpkgs.bzl", "nixpkgs_package")
+load("@bazel_tools//tools/cpp:lib_cc_configure.bzl", "get_cpu_value")
+load("@io_bazel_rules_go//go/private:platforms.bzl", "PLATFORMS")
+
+def _detect_host_platform(ctx):
```

Comment:
I take issue with how this logic is structured.

1. It hides the explicit mappings behind "ifology" statements
2. It takes the "whitelist" approach to supported architectures, thus forcing itself to be always in sync with supported golang architectures. 

I would propose to make the mappings more explicit, extendable and open - for example:

```
BAZEL_OS_TO_GO_OS = {
  "mac os x": "dawin",
}
# This allows us to remember only os names differing from official Bazel names,
# and what is more: pass-thru unexpected names
def get_go_os_name(ctx):
  bazel_os_name = ctx.os.name.lower()
  return BAZEL_OS_TO_GO_OS.get(bazel_os_name, default=bazel_os_name)

BAZEL_ARCH_TO_GO_ARCH = {
  "i686": "386",
  "aarch64": "arm64",
  "armv6l": "arm",
  "armv7l": "arm",
}
# This allows us to remember only arch names differing from official Bazel names,
# and what is more: pass-thru unexpected ones (go may support something we 
# are not yet aware of
def get_go_arch_name(ctx):
  bazel_os_name = ctx.os.name.lower()
  if bazel_os_name in ["linux", mac os x"]:
     
     res = ctx.execute(["uname", "-p"])
     if res.return_code == 0 and ("unknown" not in res.stdout):
       return BAZEL_ARCH_TO_GO_ARCH.get(res.stdout.strip(), default=res.stdout.strip())
       
     res = ctx.execute(["uname", "-m"])
     if res.return_code == 0 and ("unknown" not in res.stdout):
       return BAZEL_ARCH_TO_GO_ARCH.get(res.stdout.strip(), default=res.stdout.strip())
       
     # Pick your poison:
     # fail("Unable to run uname {-p|-m} to detect os arch")
     # return "amd64"
  else:
  # Else assume amd64 because reasons
    return "amd64"
    
    
def _detect_host_platform(ctx):
  get_go_os_name(), get_go_arch_name()
```